File: src/main/java/mediathek/gui/dialogEinstellungen/PanelProgrammPfade.java
package mediathek.gui.dialogEinstellungen;

import mediathek.config.Konstanten;
import mediathek.config.application.ApplicationConfiguration;
import mediathek.gui.dialog.HelpTextDialog;
import mediathek.tool.GetFile;
import mediathek.tool.GuiFunktionenProgramme;
import mediathek.tool.SVGIconUtilities;
import org.apache.commons.lang3.SystemUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import javax.swing.*;
import javax.swing.border.TitledBorder;
import javax.swing.event.DocumentEvent;
import javax.swing.event.DocumentListener;
import java.awt.*;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;
import java.io.File;

public class PanelProgrammPfade extends JPanel {
    private static final Logger logger = LogManager.getLogger();
    private static final Color COLOR_PINK = new Color(255, 200, 200);
    private final boolean vlc, ffmpeg;
    private final JFrame parentComponent;

    public PanelProgrammPfade(JFrame parentFrame, boolean vvlc, boolean fffmpeg) {
        initComponents();
        vlc = vvlc;
        ffmpeg = fffmpeg;
        parentComponent = parentFrame;
        init();
        initBeob();
    }

    private void init() {
        jButtonVlcPfad.setIcon(SVGIconUtilities.createSVGIcon("icons/fontawesome/folder-open.svg"));
        jButtonFFmpegPfad.setIcon(SVGIconUtilities.createSVGIcon("icons/fontawesome/folder-open.svg"));
        jButtonHilfe.setIcon(SVGIconUtilities.createSVGIcon("icons/fontawesome/circle-question.svg"));
        jPanelVlc.setVisible(vlc);

        jPanelFFmpeg.setVisible(ffmpeg);
        var applicationConfiguration = ApplicationConfiguration.getInstance();
        if (applicationConfiguration.getStandardVlcPath().isEmpty()) {
            applicationConfiguration.setStandardVlcPath(GuiFunktionenProgramme.getMusterPfadVlc());
        }
        if (applicationConfiguration.getStandardFFmpegPath().isEmpty()) {
            applicationConfiguration.setStandardFFmpegPath(GuiFunktionenProgramme.getMusterPfadFFmpeg());
        }
        jTextFieldVlc.setText(applicationConfiguration.getStandardVlcPath());
        jTextFieldFFmpeg.setText(applicationConfiguration.getStandardFFmpegPath());
    }

    private void initBeob() {
        jTextFieldVlc.getDocument().addDocumentListener(new BeobDoc());
        jTextFieldFFmpeg.getDocument().addDocumentListener(new BeobDoc());

        jButtonVlcPfad.addActionListener(new BeobPfad(jTextFieldVlc));
        jButtonFFmpegPfad.addActionListener(new BeobPfad(jTextFieldFFmpeg));
        jButtonVlcSuchen.addActionListener(_ -> {
            ApplicationConfiguration.getInstance().setStandardVlcPath("");
            jTextFieldVlc.setText(GuiFunktionenProgramme.getMusterPfadVlc());
        });

        jButtonFFmpegSuchen.addActionListener(_ -> {
            ApplicationConfiguration.getInstance().setStandardFFmpegPath("");
            jTextFieldFFmpeg.setText(GuiFunktionenProgramme.getMusterPfadFFmpeg());
        });
        jButtonHilfe.addActionListener(_ -> HelpTextDialog.show(parentComponent, GetFile.getHilfeSuchen(Konstanten.PFAD_HILFETEXT_STANDARD_PSET)));
    }

    private void check() {
        var applicationConfiguration = ApplicationConfiguration.getInstance();
        applicationConfiguration.setStandardVlcPath(jTextFieldVlc.getText());
        applicationConfiguration.setStandardFFmpegPath(jTextFieldFFmpeg.getText());

        try {
            if (jTextFieldVlc.getText().isEmpty()) {
                jTextFieldVlc.setBackground(COLOR_PINK);
            }
            else if (!new File(applicationConfiguration.getStandardVlcPath()).exists()) {
                jTextFieldVlc.setBackground(COLOR_PINK);
            }
            else {
                jTextFieldVlc.setBackground(javax.swing.UIManager.getDefaults().getColor("TextField.background"));
            }
        }
        catch (Exception ex) {
            jTextFieldVlc.setBackground(COLOR_PINK);
        }

        try {
            if (jTextFieldFFmpeg.getText().isEmpty()) {
                jTextFieldFFmpeg.setBackground(COLOR_PINK);
            }
            else if (!new File(applicationConfiguration.getStandardFFmpegPath()).exists()) {
                jTextFieldFFmpeg.setBackground(COLOR_PINK);
            }
            else {
                jTextFieldFFmpeg.setBackground(javax.swing.UIManager.getDefaults().getColor("TextField.background"));
            }
        }
        catch (Exception ex) {
            jTextFieldFFmpeg.setBackground(COLOR_PINK);
        }
    }

    private class BeobDoc implements DocumentListener {

        @Override
        public void insertUpdate(DocumentEvent e) {
            check();
        }

        @Override
        public void removeUpdate(DocumentEvent e) {
            check();
        }

        @Override
        public void changedUpdate(DocumentEvent e) {
            check();
        }
    }

    private class BeobPfad implements ActionListener {

        private final JTextField textField;

        public BeobPfad(JTextField ttextField) {
            textField = ttextField;
        }

        @Override
        public void actionPerformed(ActionEvent e) {
            //we can use native chooser on Mac...
            if (SystemUtils.IS_OS_MAC_OSX) {
                FileDialog chooser = new FileDialog(parentComponent, "Programmdatei auswählen");
                chooser.setMode(FileDialog.LOAD);
                chooser.setVisible(true);
                if (chooser.getFile() != null) {
                    try {
                        textField.setText(new File(chooser.getDirectory() + chooser.getFile()).getAbsolutePath());
                    }
                    catch (Exception ex) {
                        logger.error(ex);
                    }
                }
            }
            else {
                int returnVal;
                JFileChooser chooser = new JFileChooser();
                chooser.setFileSelectionMode(JFileChooser.FILES_ONLY);
                chooser.setFileHidingEnabled(false);
                if (textField.getText().isEmpty()) {
                    chooser.setCurrentDirectory(new File(SystemUtils.USER_HOME));
                }
                else {
                    chooser.setCurrentDirectory(new File(textField.getText()));
                }
                returnVal = chooser.showOpenDialog(null);
                if (returnVal == JFileChooser.APPROVE_OPTION) {
                    try {
                        textField.setText(chooser.getSelectedFile().getAbsolutePath());
                    }
                    catch (Exception ex) {
                        logger.error(ex);
                    }
                }
            }
        }
    }

    /** This method is called from within the constructor to
     * initialize the form.
     * WARNING: Do NOT modify this code. The content of this method is
     * always regenerated by the Form Editor.
     */
    // <editor-fold defaultstate="collapsed" desc="Generated Code">//GEN-BEGIN:initComponents
    // Generated using JFormDesigner non-commercial license
    private void initComponents() {
        var jScrollPane1 = new JScrollPane();
        var jPanel1 = new JPanel();
        jPanelVlc = new JPanel();
        jTextFieldVlc = new JTextField();
        jButtonVlcPfad = new JButton();
        jButtonVlcSuchen = new JButton();
        jButtonHilfe = new JButton();
        jPanelFFmpeg = new JPanel();
        jTextFieldFFmpeg = new JTextField();
        jButtonFFmpegSuchen = new JButton();
        jButtonFFmpegPfad = new JButton();

        //======== this ========

        //======== jScrollPane1 ========
        {

            //======== jPanel1 ========
            {

                //======== jPanelVlc ========
                {
                    jPanelVlc.setBorder(new TitledBorder("Pfad zum VLC-Player ausw\u00e4hlen"));

                    //---- jButtonVlcPfad ----
                    jButtonVlcPfad.setIcon(new ImageIcon(getClass().getResource("/mediathek/res/muster/button-file-open.png")));
                    jButtonVlcPfad.setToolTipText("Programm ausw\u00e4hlen");

                    //---- jButtonVlcSuchen ----
                    jButtonVlcSuchen.setText("suchen");
                    jButtonVlcSuchen.setToolTipText("nach dem Programm suchen");

                    GroupLayout jPanelVlcLayout = new GroupLayout(jPanelVlc);
                    jPanelVlc.setLayout(jPanelVlcLayout);
                    jPanelVlcLayout.setHorizontalGroup(
                        jPanelVlcLayout.createParallelGroup()
                            .addGroup(jPanelVlcLayout.createSequentialGroup()
                                .addContainerGap()
                                .addComponent(jTextFieldVlc)
                                .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                                .addComponent(jButtonVlcPfad)
                                .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                                .addComponent(jButtonVlcSuchen)
                                .addContainerGap())
                    );
                    jPanelVlcLayout.setVerticalGroup(
                        jPanelVlcLayout.createParallelGroup()
                            .addGroup(jPanelVlcLayout.createSequentialGroup()
                                .addContainerGap()
                                .addGroup(jPanelVlcLayout.createParallelGroup(GroupLayout.Alignment.CENTER)
                                    .addComponent(jTextFieldVlc, GroupLayout.PREFERRED_SIZE, GroupLayout.DEFAULT_SIZE, GroupLayout.PREFERRED_SIZE)
                                    .addComponent(jButtonVlcPfad)
                                    .addComponent(jButtonVlcSuchen))
                                .addContainerGap())
                    );
                    jPanelVlcLayout.linkSize(SwingConstants.VERTICAL, new Component[] {jButtonVlcPfad, jTextFieldVlc});
                }

                //---- jButtonHilfe ----
                jButtonHilfe.setIcon(new ImageIcon(getClass().getResource("/mediathek/res/muster/button-help.png")));
                jButtonHilfe.setToolTipText("Hilfe anzeigen");

                //======== jPanelFFmpeg ========
                {
                    jPanelFFmpeg.setBorder(new TitledBorder("Pfad zu ffmpeg ausw\u00e4hlen"));

                    //---- jButtonFFmpegSuchen ----
                    jButtonFFmpegSuchen.setText("suchen");
                    jButtonFFmpegSuchen.setToolTipText("nach dem Programm suchen");

                    //---- jButtonFFmpegPfad ----
                    jButtonFFmpegPfad.setIcon(new ImageIcon(getClass().getResource("/mediathek/res/muster/button-file-open.png")));
                    jButtonFFmpegPfad.setToolTipText("Programm ausw\u00e4hlen");

                    GroupLayout jPanelFFmpegLayout = new GroupLayout(jPanelFFmpeg);
                    jPanelFFmpeg.setLayout(jPanelFFmpegLayout);
                    jPanelFFmpegLayout.setHorizontalGroup(
                        jPanelFFmpegLayout.createParallelGroup()
                            .addGroup(jPanelFFmpegLayout.createSequentialGroup()
                                .addContainerGap()
                                .addComponent(jTextFieldFFmpeg, GroupLayout.DEFAULT_SIZE, 345, Short.MAX_VALUE)
                                .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                                .addComponent(jButtonFFmpegPfad)
                                .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                                .addComponent(jButtonFFmpegSuchen)
                                .addContainerGap())
                    );
                    jPanelFFmpegLayout.setVerticalGroup(
                        jPanelFFmpegLayout.createParallelGroup()
                            .addGroup(jPanelFFmpegLayout.createSequentialGroup()
                                .addContainerGap()
                                .addGroup(jPanelFFmpegLayout.createParallelGroup(GroupLayout.Alignment.CENTER)
                                    .addComponent(jTextFieldFFmpeg, GroupLayout.PREFERRED_SIZE, GroupLayout.DEFAULT_SIZE, GroupLayout.PREFERRED_SIZE)
                                    .addComponent(jButtonFFmpegPfad)
                                    .addComponent(jButtonFFmpegSuchen))
                                .addContainerGap(GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                    );
                    jPanelFFmpegLayout.linkSize(SwingConstants.VERTICAL, new Component[] {jButtonFFmpegPfad, jButtonFFmpegSuchen, jTextFieldFFmpeg});
                }

                GroupLayout jPanel1Layout = new GroupLayout(jPanel1);
                jPanel1.setLayout(jPanel1Layout);
                jPanel1Layout.setHorizontalGroup(
                    jPanel1Layout.createParallelGroup()
                        .addGroup(jPanel1Layout.createSequentialGroup()
                            .addContainerGap()
                            .addGroup(jPanel1Layout.createParallelGroup()
                                .addComponent(jPanelVlc, GroupLayout.DEFAULT_SIZE, GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE)
                                .addGroup(GroupLayout.Alignment.TRAILING, jPanel1Layout.createSequentialGroup()
                                    .addGap(0, 0, Short.MAX_VALUE)
                                    .addComponent(jButtonHilfe))
                                .addComponent(jPanelFFmpeg, GroupLayout.DEFAULT_SIZE, GroupLayout.DEFAULT_SIZE, Short.MAX_VALUE))
                            .addContainerGap())
                );
                jPanel1Layout.setVerticalGroup(
                    jPanel1Layout.createParallelGroup()
                        .addGroup(jPanel1Layout.createSequentialGroup()
                            .addContainerGap()
                            .addComponent(jPanelVlc, GroupLayout.PREFERRED_SIZE, GroupLayout.DEFAULT_SIZE, GroupLayout.PREFERRED_SIZE)
                            .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                            .addComponent(jPanelFFmpeg, GroupLayout.PREFERRED_SIZE, GroupLayout.DEFAULT_SIZE, GroupLayout.PREFERRED_SIZE)
                            .addPreferredGap(LayoutStyle.ComponentPlacement.RELATED)
                            .addComponent(jButtonHilfe)
                            .addContainerGap(84, Short.MAX_VALUE))
                );
            }
            jScrollPane1.setViewportView(jPanel1);
        }

        GroupLayout layout = new GroupLayout(this);
        setLayout(layout);
        layout.setHorizontalGroup(
            layout.createParallelGroup()
                .addGroup(layout.createSequentialGroup()
                    .addContainerGap()
                    .addComponent(jScrollPane1)
                    .addContainerGap())
        );
        layout.setVerticalGroup(
            layout.createParallelGroup()
                .addGroup(layout.createSequentialGroup()
                    .addContainerGap()
                    .addComponent(jScrollPane1)
                    .addContainerGap())
        );
    }// </editor-fold>//GEN-END:initComponents

    // Variables declaration - do not modify//GEN-BEGIN:variables
    // Generated using JFormDesigner non-commercial license
    private JPanel jPanelVlc;
    private JTextField jTextFieldVlc;
    private JButton jButtonVlcPfad;
    private JButton jButtonVlcSuchen;
    private JButton jButtonHilfe;
    private JPanel jPanelFFmpeg;
    private JTextField jTextFieldFFmpeg;
    private JButton jButtonFFmpegSuchen;
    private JButton jButtonFFmpegPfad;
    // End of variables declaration//GEN-END:variables
}

File: src/main/kotlin/mediathek/gui/dialogEinstellungen/PanelProgrammPfade.kt
/*
 * Copyright (c) 2026 derreisende77.
 * This code was developed as part of the MediathekView project https://github.com/mediathekview/MediathekView
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package mediathek.gui.dialogEinstellungen

import mediathek.config.Konstanten
import mediathek.config.application.ApplicationConfiguration
import mediathek.gui.dialog.HelpTextDialog
import mediathek.tool.GetFile
import mediathek.tool.GuiFunktionenProgramme
import mediathek.tool.SVGIconUtilities
import org.apache.commons.lang3.SystemUtils
import org.apache.logging.log4j.LogManager
import java.awt.Color
import java.awt.FileDialog
import java.io.File
import javax.swing.JFileChooser
import javax.swing.JFrame
import javax.swing.JTextField
import javax.swing.UIManager
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener

class PanelProgrammPfade(
    private val parentComponent: JFrame?,
    private val showVlcSettings: Boolean,
    private val showFFmpegSettings: Boolean,
) : PanelProgrammPfadeBase() {
    init {
        configureComponentMetadata()
        initializePanel()
        installListeners()
    }

    private fun configureComponentMetadata() {
        jPanelVlc.name = PanelProgrammPfadeComponentNames.VLC_PANEL
        jTextFieldVlc.name = PanelProgrammPfadeComponentNames.VLC_PATH
        jButtonVlcPfad.name = PanelProgrammPfadeComponentNames.CHOOSE_VLC
        jButtonVlcSuchen.name = PanelProgrammPfadeComponentNames.SEARCH_VLC
        jButtonHilfe.name = PanelProgrammPfadeComponentNames.HELP
        jPanelFFmpeg.name = PanelProgrammPfadeComponentNames.FFMPEG_PANEL
        jTextFieldFFmpeg.name = PanelProgrammPfadeComponentNames.FFMPEG_PATH
        jButtonFFmpegSuchen.name = PanelProgrammPfadeComponentNames.SEARCH_FFMPEG
        jButtonFFmpegPfad.name = PanelProgrammPfadeComponentNames.CHOOSE_FFMPEG
    }

    private fun initializePanel() {
        jButtonVlcPfad.icon = SVGIconUtilities.createSVGIcon("icons/fontawesome/folder-open.svg")
        jButtonFFmpegPfad.icon = SVGIconUtilities.createSVGIcon("icons/fontawesome/folder-open.svg")
        jButtonHilfe.icon = SVGIconUtilities.createSVGIcon("icons/fontawesome/circle-question.svg")
        jPanelVlc.isVisible = showVlcSettings
        jPanelFFmpeg.isVisible = showFFmpegSettings

        val config = ApplicationConfiguration.getInstance()
        if (config.standardVlcPath.isEmpty()) {
            config.standardVlcPath = GuiFunktionenProgramme.getMusterPfadVlc()
        }
        if (config.standardFFmpegPath.isEmpty()) {
            config.standardFFmpegPath = GuiFunktionenProgramme.getMusterPfadFFmpeg()
        }
        jTextFieldVlc.text = config.standardVlcPath
        jTextFieldFFmpeg.text = config.standardFFmpegPath
    }

    private fun installListeners() {
        jTextFieldVlc.document.addDocumentListener(ProgramPathDocumentListener())
        jTextFieldFFmpeg.document.addDocumentListener(ProgramPathDocumentListener())

        jButtonVlcPfad.addActionListener { chooseProgramFile(jTextFieldVlc) }
        jButtonFFmpegPfad.addActionListener { chooseProgramFile(jTextFieldFFmpeg) }
        jButtonVlcSuchen.addActionListener {
            ApplicationConfiguration.getInstance().standardVlcPath = ""
            jTextFieldVlc.text = GuiFunktionenProgramme.getMusterPfadVlc()
        }
        jButtonFFmpegSuchen.addActionListener {
            ApplicationConfiguration.getInstance().standardFFmpegPath = ""
            jTextFieldFFmpeg.text = GuiFunktionenProgramme.getMusterPfadFFmpeg()
        }
        jButtonHilfe.addActionListener {
            HelpTextDialog.show(
                parentComponent,
                GetFile.getHilfeSuchen(Konstanten.PFAD_HILFETEXT_STANDARD_PSET),
            )
        }
    }

    private fun checkPaths() {
        val config = ApplicationConfiguration.getInstance()
        config.standardVlcPath = jTextFieldVlc.text
        config.standardFFmpegPath = jTextFieldFFmpeg.text
        updatePathValidation(jTextFieldVlc, config.standardVlcPath)
        updatePathValidation(jTextFieldFFmpeg, config.standardFFmpegPath)
    }

    private fun updatePathValidation(textField: JTextField, configuredPath: String) {
        textField.background = try {
            if (textField.text.isEmpty() || !File(configuredPath).exists()) {
                INVALID_PATH_COLOR
            } else {
                UIManager.getDefaults().getColor(TEXT_FIELD_BACKGROUND_KEY)
            }
        } catch (_: Exception) {
            INVALID_PATH_COLOR
        }
    }

    private fun chooseProgramFile(textField: JTextField) {
        if (SystemUtils.IS_OS_MAC_OSX) {
            chooseProgramFileWithNativeDialog(textField)
        } else {
            chooseProgramFileWithSwingDialog(textField)
        }
    }

    private fun chooseProgramFileWithNativeDialog(textField: JTextField) {
        val chooser = FileDialog(parentComponent, PROGRAM_FILE_DIALOG_TITLE).apply {
            mode = FileDialog.LOAD
            isVisible = true
        }
        val selectedFile = chooser.file ?: return
        try {
            textField.text = File(chooser.directory + selectedFile).absolutePath
        } catch (ex: Exception) {
            logger.error(ex)
        }
    }

    private fun chooseProgramFileWithSwingDialog(textField: JTextField) {
        val chooser = JFileChooser().apply {
            fileSelectionMode = JFileChooser.FILES_ONLY
            isFileHidingEnabled = false
            currentDirectory = if (textField.text.isEmpty()) {
                File(SystemUtils.USER_HOME)
            } else {
                File(textField.text)
            }
        }
        if (chooser.showOpenDialog(null) == JFileChooser.APPROVE_OPTION) {
            try {
                textField.text = chooser.selectedFile.absolutePath
            } catch (ex: Exception) {
                logger.error(ex)
            }
        }
    }

    private inner class ProgramPathDocumentListener : DocumentListener {
        override fun insertUpdate(event: DocumentEvent) = checkPaths()

        override fun removeUpdate(event: DocumentEvent) = checkPaths()

        override fun changedUpdate(event: DocumentEvent) = checkPaths()
    }

    private companion object {
        private val logger = LogManager.getLogger(PanelProgrammPfade::class.java)
        private val INVALID_PATH_COLOR = Color(255, 200, 200)
        private const val PROGRAM_FILE_DIALOG_TITLE = "Programmdatei auswählen"
        private const val TEXT_FIELD_BACKGROUND_KEY = "TextField.background"
    }
}

internal object PanelProgrammPfadeComponentNames {
    const val VLC_PANEL = "PanelProgrammPfade.vlcPanel"
    const val VLC_PATH = "PanelProgrammPfade.vlcPath"
    const val CHOOSE_VLC = "PanelProgrammPfade.chooseVlc"
    const val SEARCH_VLC = "PanelProgrammPfade.searchVlc"
    const val HELP = "PanelProgrammPfade.help"
    const val FFMPEG_PANEL = "PanelProgrammPfade.ffmpegPanel"
    const val FFMPEG_PATH = "PanelProgrammPfade.ffmpegPath"
    const val SEARCH_FFMPEG = "PanelProgrammPfade.searchFFmpeg"
    const val CHOOSE_FFMPEG = "PanelProgrammPfade.chooseFFmpeg"
}
